src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -356,26 +356,37 @@
   environment-module-response! :- ringutils/RingResponse
   "Process the environment module information, returning a ring response to be
   propagated back up to the caller of the environment_modules endpoint."
-  [info-from-jruby :- List
-   environment :- schema/Str]
-  (let [info-as-json (module-info-from-jruby->module-info-for-json
-                       info-from-jruby environment)]
-    (middleware-utils/json-response 200 info-as-json)))
+  ([info-from-jruby :- List]
+   (let [return-data (map #(module-info-from-jruby->module-info-for-json
+                             (first (vals %))
+                             (name (first (keys %))))
+                          (sort-nested-info-maps info-from-jruby))]
+     (middleware-utils/json-response 404 return-data)))
+  ([info-from-jruby :- List
+    environment :- schema/Str]
+   (let [info-as-json (module-info-from-jruby->module-info-for-json
+                        info-from-jruby environment)]
+     (middleware-utils/json-response 200 info-as-json))))
 
 (schema/defn ^:always-validate
   environment-module-info-fn :- IFn
   "Middleware function for constructing a Ring response from an incoming
   request for environment_modules information."
   [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
   (fn [request]
-    (let [environment (jruby-request/get-environment-from-request request)]
+    (if-let [environment (jruby-request/get-environment-from-request request)]
       (if-let [module-info
                (jruby-protocol/get-environment-module-info jruby-service
                                                            (:jruby-instance request)
                                                            environment)]
         (environment-module-response! module-info
                                       environment)
-        (rr/not-found (i18n/tru "Could not find environment ''{0}''" environment))))))
+        (rr/not-found (i18n/tru "Could not find environment ''{0}''" environment)))
+      (if-let [module-info
+               (jruby-protocol/get-all-environment-module-info jruby-service
+                                                           (:jruby-instance request))]
+        (environment-module-response! module-info)
+        (rr/not-found (i18n/tru "Could not find any puppet module data"))))))
```

Comment:
Would you expect this line to ever be hit?  If the expected behavior for when there are no modules that an empty hash be returned from the Ruby layer, I'd probably just avoid doing the extra `if-let` with "else" conditional here.